### PR TITLE
Add support for importing Agreements

### LINKED ISSUE #3: Migrate FPCA details
Now we have agreements in freeipa-fas, we need to migrate the FPCA over to ipa. and make everyone that has previously signed the agreement, signed in the new system

### PR DESCRIPTION
**Pull request by ryanlerch**
Resolves: #3

Signed-off-by: Ryan Lerch <rlerch@redhat.com>

### CODE REVIEW

File: settings_default.py [hunk 1]
```diff
@@ -1,15 +1,15 @@
 from getpass import getpass
```
Comment (abompard) [inline settings_default.py]: Sorry, I forgot to remove that file, we don't use it anymore. The config is in `config.toml.example` and `config.toml`. Can you add an `agreements` section there?

File: fas2ipa/cli.py [hunk 2]
```diff
@@ -171,6 +171,18 @@ def migrate_group(group, ipa):
     )
     try:
         ipa.group_add(name, **group_args)
+        if config.get("agreements"):
+            for agreement, agreement_details in config.get("agreements").items():
```
Comment (abompard) [inline fas2ipa/cli.py]: It looks like you never use the `agreement` variable, your dictionary's key.

File: fas2ipa/cli.py [hunk 3]
```diff
@@ -171,6 +171,18 @@ def migrate_group(group, ipa):
     )
     try:
         ipa.group_add(name, **group_args)
+        if config.get("agreements"):
```
Comment (abompard) [inline fas2ipa/cli.py]: You don't need that test, if the dict is empty the iteration over `.items()` will just do nothing.

**Review by abompard — CHANGES_REQUESTED** [review]

File: settings_default.py (hunk 1 above)
Comment (ryanlerch) [reply settings_default.py]: hah. yeah i updated my local config.toml, but forgot to put that in the example one. and my black run updated this file lol. Its deleted now tho

File: config.toml.example [hunk 4]
```diff
@@ -23,6 +23,206 @@ url = "https://admin.fedoraproject.org/accounts"
...
+    [agreements.fpca]
+    name = "Fedora Project Contributor Agreement"
+    # fas_group_prerequisites is a list of FAS Group IDs that
+    # if a fas group has this id listed as a prerequisite,
+    # we add this group to this agreement.
+    # 100002 == cla_done
+    # 100300 == packager
+    fas_group_prerequisites = [100002, 100300]
+    fas_user_group = "cla_fpca"
+    description = """
```
Comment (abompard) [inline config.toml.example]: Maybe a path to a local file instead of dumping all that in the config?

File: fas2ipa/cli.py [hunk 5]
```diff
@@ -171,6 +171,14 @@ def migrate_group(group, ipa):
     )
     try:
         ipa.group_add(name, **group_args)
+        for agreement_details in config.get("agreements").values():
```
Comment (abompard) [inline fas2ipa/cli.py]: In toml, you represent a list of dicts as such:
```
[[agreements]]
name = "agreement 1"
key = "value"

[[agreements]]
name = "agreement 2"
key = "value"
```
(note the double brackets)
This seems to be a better data structure for what you want to do.

File: fas2ipa/cli.py [hunk 6]
```diff
@@ -259,6 +267,14 @@ def migrate_user(config, person, ipa):
         )
         try:
             ipa.user_add(person["username"], **user_args)
+            for group in person["memberships"]:
+                for agreement, agreement_details in config.get("agreements").items():
```
Comment (abompard) [inline fas2ipa/cli.py]: Same remark here.

**Review by abompard — CHANGES_REQUESTED** [review]
I'm not so sure about the logic here.

If I understand correctly, in FAS there's group trees. And one of the roots is `cla_done`. Other groups can mark it as prerequisite, and then other groups can mark those other groups as prerequisite, which means that they will transitively require `cla_done`.

With our flat system of agreements, I think we should first flatten the tree instead of trying to list all prerequisites, because we may forget some. Espcially since we're not currently working with production data, so we can't aim to be exhaustive now.

I would recommend the following logic:
- list the name of the groups for each agreements in the configuration, for example `cla_done`.
- get all groups, as we do now

For each agreement's corresponding group name, do the following:
1. get the group id for that group name
2. prepare a list of groups IDs that will require that group as a prerequisite, and pre-fill it with only that group ID
3. recursively add to that list the IDs of the groups that have flagged as prerequisite a group id that is already in the list. Stop the recursion when there is nothing to add.

You now have an exhaustive list of all the group ids that require this argument, directly of transitively. The list also contains the id of the original group, but it's not really a problem since we don't migrate it.
When migrating groups, you can check that their id is in the list and add them to the corresponding agreement.

Does it make sense?

File: fas2ipa/cli.py [hunk 7]
```diff
@@ -358,6 +370,54 @@ def add_users_to_groups(config, instances, groups_to_users, category):
...
+                "fasagreement_add",
+                agreement["name"],
+                {"description": agreement_description},
+            )
+            print_status(Status.ADDED, f"Added {agreement['name']}")
+        except python_freeipa.exceptions.DuplicateEntry as e:
+            click.echo(e)
+
+        toplevel_prereq = fas.send_request(
+            "/group/list/",
```
Comment (abompard) [inline fas2ipa/cli.py]: It would be more coherent with the other API calls if this did not have the trailing slash.

File: fas2ipa/cli.py [hunk 8]
```diff
@@ -358,6 +370,54 @@ def add_users_to_groups(config, instances, groups_to_users, category):
...
+    for agreement in config.get("agreement"):
+        with open(agreement["description"], "r") as f:
+            agreement_description = f.read()
+        try:
+            ipa._request(
+                "fasagreement_add",
+                agreement["name"],
+                {"description": agreement_description},
+            )
+            print_status(Status.ADDED, f"Added {agreement['name']}")
```
Comment (abompard) [inline fas2ipa/cli.py]: The API call actually gives you back an indication on whether it did something, in the result's `completed` value. I propose the following change:
```diff
@@ -412,10 +412,17 @@ def migrate_agreements(config, fas, ipa):
         )["groups"][0]["id"]
         agreement_required = find_requirements(fas_groups["groups"], toplevel_prereq)
         for dep_name in agreement_required:
-            ipa._request(
+            result = ipa._request(
                 "fasagreement_add_group", agreement["name"], {"group": dep_name}
             )
-            print_status(Status.ADDED, f"Added {dep_name} to the {agreement['name']}")
+            if result["completed"]:
+                print_status(
+                    Status.ADDED, f"Added {dep_name} to the {agreement['name']}"
+                )
+            else:
+                print_status(
+                    Status.SKIPPED, f"{dep_name} already requires {agreement['name']}"
+                )
```

File: fas2ipa/cli.py [hunk 9]
```diff
@@ -400,11 +460,15 @@ def cli(skip_groups, only_members, users_start_at):
...
 
+    if config.get("agreement"):
+        migrate_agreements(config, fas, ipa)
+
     alphabet = list(string.ascii_lowercase)
     if users_start_at:
         start_index = alphabet.index(users_start_at.lower())
         del alphabet[:start_index]
 
+
```
Comment (abompard) [inline fas2ipa/cli.py]: I think that will make flake8 unhappy

File: config.toml.example [hunk 10]
```diff
@@ -26,3 +26,9 @@ instances = ["ipa.fas2ipa.test"]
 cert_path = "/etc/ipa/ca.crt"
 username = "admin"
 password = "adminPassw0rd!"
+
+[[agreement]]
+name = "Fedora Project Contributor Agreement"
+group_prerequisite = "cla_done"
+signed_group = "cla_fpca"
+description = "FPCA.txt"
```
Comment (abompard) [inline config.toml.example]: Calling that `description_file` is a bit better I think

File: fas2ipa/cli.py [hunk 11]
```diff
@@ -358,6 +370,54 @@ def add_users_to_groups(config, instances, groups_to_users, category):
...
+            dependent_groups.append(group["name"])
+            subdeps = find_requirements(groups, group["id"])
+            dependent_groups.extend(subdeps)
+    return dependent_groups
+
+
+def migrate_agreements(config, fas, ipa):
+
+    click.echo("Getting the list of groups...")
+    fas_groups = fas.send_request(
```
Comment (abompard) [inline fas2ipa/cli.py]: Actually I think we should:
- first create the agreements
- when migrating the groups, add them to the agreement.

This will save us from requesting the list of groups twice. What do you think?

File: fas2ipa/cli.py [hunk 12]
```diff
@@ -296,6 +296,18 @@ def migrate_user(config, person, ipa):
...
+                for agreement in config.get("agreement"):
+                    if group["name"] == agreement["signed_group"]:
+                        ipa._request(
+                            "fasagreement_add_user",
+                            agreement["name"],
+                            {"user": person["username"]},
+                        )
+                        print_status(
+                            Status.ADDED,
+                            f"Added {person['username']} to the {agreement['name']} agreement",
```
Comment (abompard) [inline fas2ipa/cli.py]: This will probably mess up the output, because the status returned from this function is echoed at the end of an existing line.

File: fas2ipa/cli.py [hunk 13]
```diff
@@ -296,6 +296,18 @@ def migrate_user(config, person, ipa):
         )
         try:
             ipa.user_add(person["username"], **user_args)
+            for group in person["memberships"]:
```
Comment (abompard) [inline fas2ipa/cli.py]: If you put this here, it will never get executed for existing users. And we may want to update it if a user has signed an agreement between two runs.

File: fas2ipa/cli.py [hunk 14]
```diff
@@ -296,6 +296,18 @@ def migrate_user(config, person, ipa):
...
+                for agreement in config.get("agreement"):
+                    if group["name"] == agreement["signed_group"]:
```
Comment (abompard) [inline fas2ipa/cli.py]: You could avoid one level of loop nesting if you prepared the group names list as:
```python
group_names = [g["name"] for g in person["memberships"]]
```
And the you could check with:
```python
if agreement["signed_group"] in group_names:
```
Both versions would work, it's my personal preference to avoid nesting levels when possible because I find it more readable, but you may have a different point of view.

**Review by abompard — CHANGES_REQUESTED** [review]

File: fas2ipa/cli.py (hunk 13 above)
Comment (abompard) [reply fas2ipa/cli.py]: I think this comment is still valid. To verify, I delete the agreement but not the users and run the script again (you could also add another agreement that maps to `fpca_done`). The users should be marked as having signed the agreement, and currently they don't.

**Review by abompard — CHANGES_REQUESTED** [review]
One comment is still unresolved.

File: fas2ipa/cli.py (hunk 13 above)
Comment (ryanlerch) [reply fas2ipa/cli.py]: whoops! sorry i thought i fixed this, but obviously didnt...

should be fixed now.

File: fas2ipa/cli.py [hunk 15]
```diff
@@ -248,6 +286,17 @@ def migrate_users(config, users, instances):
             groups_to_member_usernames[groupname].append(person["username"])
             if membership["role_type"] in ["administrator", "sponsor"]:
                 groups_to_sponsor_usernames[groupname].append(person["username"])
+                group_names = [g["name"] for g in person["memberships"]]
```
Comment (abompard) [inline fas2ipa/cli.py]: This should not be in the `if` block, because it will be set only for sponsors and administrators, which gives a "local variable 'group_names' referenced before assignment" error on users who are just members.

**Review by abompard — CHANGES_REQUESTED** [review]

**Comment by abompard** [comment]
I've noticed that there are some users in groups that require the FPCA but are not in the `fpca_done` group, only the `cla_done` group. What do you think we should do with those? Mark them as having signed the FPCA too? (example:  `aaronhetherington`)

**Comment by ryanlerch** [comment]
> I've noticed that there are some users in groups that require the FPCA but are not in the `fpca_done` group, only the `cla_done` group. What do you think we should do with those? Mark them as having signed the FPCA too? (example: `aaronhetherington`)

this was somthing i wasnt sure of -- i only implemented the fcpa_done, didnt want to mark people as signing the fcpa when they indeed have not. My assumption here 

Maybe the best approach here is to make the config item "signed_groups = ["cla_fpca"]" to allow for multiples if we need them? 

Might have to ask someone around / involved with the CLA > FPCA transition as well about why this might be the case.

**Comment by abompard** [comment]
Yeah a list would be safer. During migration we can't add those users to groups they were members of before, because they aren't marked as having signed the FPCA, so we need to decide what to do.

**Comment by nirik** [comment]
> I've noticed that there are some users in groups that require the FPCA but are not in the `fpca_done` group, only the `cla_done` group. What do you think we should do with those? Mark them as having signed the FPCA too? (example: `aaronhetherington`)

So, I think this might be where a group was created, users were added to it, then the cla_done constraint was applied to the group. :( In that case I think we need to either forbid adding that constraint to the group (sorry, some current users don't match that critera, remove them or add them to required groups before adding this constraint) or when you add the constraint/prereq to a existing group it just removes anyone who no longer meets the requirements?

At least thats the case I can think of... can you see any others?

**Comment by ryanlerch** [comment]
> > I've noticed that there are some users in groups that require the FPCA but are not in the `fpca_done` group, only the `cla_done` group. What do you think we should do with those? Mark them as having signed the FPCA too? (example: `aaronhetherington`)
> 
> So, I think this might be where a group was created, users were added to it, then the cla_done constraint was applied to the group. :( In that case I think we need to either forbid adding that constraint to the group (sorry, some current users don't match that critera, remove them or add them to required groups before adding this constraint) or when you add the constraint/prereq to a existing group it just removes anyone who no longer meets the requirements?
> 
> At least thats the case I can think of... can you see any others?

AIUI, the issue here is on the user side, right? No groups actaully depend on cla_fcpa. Only cla_done (either by inheritance or directly). However, most users are members of both fcpa_done and cla_done (but what abompard was pointing out was that some users are only members of cla_done)

When writing the script, we add all groups that depend on cla_done to the FCPA agreement. Then only the users that are in the cla_fcpa are added to the FCPA agreement ("signing" it).

**Comment by ryanlerch** [comment]
Just added an extra commit to allow us to specify multiple fas groups. If a user is a member of one of the groups specified in "signed_groups" then they are added to that agreement.

**Comment by abompard** [comment]
I think not adding the constraint to groups that have non-matching users will not work in our situation, because that will leave out large groups like "packagers". We could add users only if they have signed the agreement and leave them out of those groups otherwise, but I've checked and it's not only old accounts, there are accounts in that situation that signed in pretty recently (2020).
Last option I can think of would be to mark them as having signed the FCPA if they have signed the old CLA. I'm not sure if that's a good move from a legal point of view. Ryan's last commit would make it simple to do that if we chose to.

**Comment by abompard** [comment]
I'll merge it as is because it works fine but I'd be interested in continuing that discussion here (or elsewhere).

**Comment by abompard** [comment]
For the record there are currently 409 users who are in the `cla_done` group but not in the `cla_fcpa` group.

**Comment by nirik** [comment]
Interesting, so are any of those in cla_fedora or cla* ?

They may be users who agreed to the old cla, but never accepted the fpca?
also, they may be convered under one of those other cla's... cla_ibm, cla_dell, cla_redhat, etc?

**Comment by abompard** [comment]
They all have `cla_done` at least, but indeed some have `cla_*` also.
